tangelo/linq/simulator.py

```diff
@@ -26,515 +26,35 @@
-        least-significant-qubit (lsq) -first order. e.g the string '100' means
-        qubit 0 measured in basis state |1>, and qubit 1 & 2 both measured in
-        state |0>.
-
-        Args:
-            statevector(list or ndarray(complex)): an iterable 1D data-structure
-                containing the amplitudes.
-
-        Returns:
-            dict: A dictionary whose keys are bitstrings representing the
-                multi-qubit states with the least significant qubit first (e.g
-                '100' means qubit 0 in state |1>, and qubit 1 and 2 in state
-                |0>), and the associated value is the corresponding frequency.
-                Unless threshold=0., this dictionary will be sparse.
-        """
-
-        n_qubits = int(math.log2(len(statevector)))
-        frequencies = dict()
-        for i, amplitude in enumerate(statevector):
-            frequency = abs(amplitude)**2
-            if (frequency - self.freq_threshold) >= 0.:
-                frequencies[self.__int_to_binstr(i, n_qubits)] = frequency
-
-        # If n_shots, has been specified, then draw that amount of samples from the distribution
-        # and return empirical frequencies instead. Otherwise, return the exact frequencies
-        if not self.n_shots:
-            return frequencies
-        else:
-            xk, pk = [], []
-            for k, v in frequencies.items():
-                xk.append(int(k[::-1], 2))
-                pk.append(frequencies[k])
-            distr = stats.rv_discrete(name='distr', values=(np.array(xk), np.array(pk)))
-
-            # Generate samples from distribution. Cut in chunks to ensure samples fit in memory, gradually accumulate
-            chunk_size = 10**7
-            n_chunks = self.n_shots // chunk_size
-            freqs_shots = Counter()
-
-            for i in range(n_chunks+1):
-                this_chunk = self.n_shots % chunk_size if i == n_chunks else chunk_size
-                samples = distr.rvs(size=this_chunk)
-                freqs_shots += Counter(samples)
-            freqs_shots = {self.__int_to_binstr_lsq(k, n_qubits): v / self.n_shots for k, v in freqs_shots.items()}
-            return freqs_shots
 
-    def __int_to_binstr(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the order
-        specified for the state vector.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1] if (self.statevector_order == "msq_first") else state_binstr
 
-    def __int_to_binstr_lsq(self, i, n_qubits):
-        """Convert an integer into a bit string of size n_qubits, in the
-        least-significant qubit order.
-        """
-        bs = bin(i).split('b')[-1]
-        state_binstr = "0" * (n_qubits - len(bs)) + bs
-        return state_binstr[::-1]
+def Simulator(target=default_simulator, n_shots=None, noise_model=None, **kwargs) -> SimulatorBase:
```

Comment:
I did not realize, but this is a function wrapper that returns a specialized class that depends on the parameters passed to it.

Although it is a function, you kept the capital S, I assume in an effort to break nothing, but also because functionally this behaves like an object constructor (factory).

The other way to achieve this would be to make `Simulator` a class, and then return the appropriate object by throwing the body of your function into the `__new__` method and leaving `__init__` as `pass` (the latter effectively means that you simply will never instantiate `Simulate`, but always another class). A very little example:

```
# Pretend QiskitSimulator has been defined and inherits from SimulatorBase
class Simulator():
    def __new__(self, target):
        if target == 'qiskit':
            return QiskitSimulator()
    def __init__(self):
        pass

qiskit_sim = Simulator(target="qiskit")
print(type(qiskit_sim)) # this is indeed a QiskitSimulator object
```

My unease with your function wrapper:
- you claim it returns a `SimulatorBase` object and... well that's a bit misleading or unhelpful here. It returns a specialization of that class, which implements the method `simulate_circuit` so that's not quite the same.
- `Simulate` is a function masquerading as a class (name, return values), it is misleading as many users will rely on name conventions, but strictly adhering to name conventions will break retro-compatibility.